L1Trigger/L1TGlobal/src/GlobalBoard.cc

```diff
@@ -1168,35 +1167,51 @@ void l1t::GlobalBoard::printGmtData(const int iBxInEvent) const {
   LogTrace("L1TGlobal") << std::endl;
 }
 
-//initializer prescale counter using a semi-random value between [1, prescale value]
-const std::vector<double> l1t::GlobalBoard::semirandomNumber(const edm::Event& iEvent,
-                                                             const std::vector<double>& prescaleFactorsAlgoTrig) {
-  auto out = prescaleFactorsAlgoTrig;
+// initializer prescale counters using a semi-random value between [0, prescale value * 10 ^ precision - 1]
+const std::vector<l1t::GlobalBoard::PrescaleCounter> l1t::GlobalBoard::semirandomNumber(
+    const edm::Event& iEvent, const std::vector<double>& prescaleFactorsAlgoTrig) {
   // pick a random number from a combination of run, lumi, event numbers
-  std::srand(iEvent.id().run());
-  std::srand(std::rand() + iEvent.id().luminosityBlock());
   // this causes different (semi)random number number for different streams
   // reminder: different streams have different initial event number
+  std::srand(iEvent.id().run());
+  std::srand(std::rand() + iEvent.id().luminosityBlock());
   std::srand(std::rand() + iEvent.id().event());
+
   // very large (semi)random number
-  double const semirandom = std::rand();
-  for (auto& ps : out) {
-    // if the ps is smaller than 1 (e.g. ps=0, ps=1), it is not changed
-    // else, replace ps with a semirandom integer in the [1,ps] range
-    if (ps > 1) {
-      auto nps = semirandom - floor(semirandom / ps) * ps;
-      // if nps=0 or a wrong value (<0,>ps) use PS value (standard method)
-      if (nps > 0 and nps <= ps)
-        ps = nps;
-      else {
-        if (nps != 0)  // complain only if nps <0 or nps >PS
-          edm::LogWarning("L1TGlobal::semirandomNumber")
-              << "\n The inital prescale counter obtained by L1TGlobal::semirandomNumber is wrong."
-              << "\n This is probably do to the floating-point precision. Using the PS value."
-              << "\n semirandom = " << semirandom << "\n PS = " << ps << "\n nps = " << nps
-              << " <-- it should be in the range [0 , " << ps << "]" << std::endl;
-      }
-    }
+  std::vector<PrescaleCounter> out;
+  out.reserve(prescaleFactorsAlgoTrig.size());
+
+  for (size_t iAlgo = 0; iAlgo < prescaleFactorsAlgoTrig.size(); iAlgo++) {
+    // for prescaled triggers, initialise trigger_counter to
+    // a semirandom integer between 0 and prescale_count - 1 (inclusive)
+    out.emplace_back(PrescaleCounter(prescaleFactorsAlgoTrig[iAlgo]));
+
+    auto& prescaleCounter = out.back();
+    prescaleCounter.trigger_counter = std::rand() % prescaleCounter.prescale_count;
```

Comment:
```suggestion
    out.emplace_back(prescaleFactorsAlgoTrig[iAlgo]);
    // for prescaled triggers, initialise trigger_counter to
    // a semirandom integer between 0 and prescale_count-1 (both inclusive)
    auto& pCount = out.back();
    if (pCount.prescale_count != m_singlestep) {
      pCount.trigger_counter = rand_num % pCount.prescale_count;
    }
```

Unprescaled triggers should not be changed by `semirandomNumber` (see the original implementation and https://github.com/missirol/cmssw/commit/a6012b8bbb428aa261835d9e3a0e40bb42970cb9).

> Please note that the change actually affects the logic slightly since now there is a new random number for each seed, rather than one per function call.

Right, then let's avoid extra changes and move the call out of the loop.

Reply:
Okay, iiuc, changing trigger_counter to a value between 0 and 99 for unprescaled triggers does not really change their behaviour (they remain unprescaled). If that's the case, the if statement could be dropped.

Reply:
In addition, `pCount.trigger_counter = rand_num % pCount.prescale_count;` won't work if `pCount.prescale_count == 0`, so please use
```
if (pCount.prescale_count > 0) {
  pCount.trigger_counter = rand_num % pCount.prescale_count;
}
```cpp

Reply:
Yeah that's right, having an offset to the initial trigger_counter won't affect the behavior of unprescaled triggers. I'll fix the mod 0 case.